improves macho loader

gitoleg opened the pull request:
Previously, only little endian binaries were expected in our macho loader, so this caused unexpected behavior with big endian binaries. This PR fix it, as well adds some more guards for reliability.

Also, this PR fixes problem with an entry point for old macho binaries: they just don't contain so-called `LC_MAIN` command, and entry point as part of the binary structure is absent.
But we can workaround this problem by emitting entry point as an address of a first executable section.

ivg commented on lib/bap_llvm/llvm_macho_loader.hpp:
```diff
@@ -152,6 +137,38 @@ MACHO_SECTION_FIELD(reserved2, uint32_t)
+            return;
+        }
+    }
+    std::vector<uint64_t> addrs;
```
Big binaries can have quite a few section, so I don't think that we should use such inefficient algorithm. You can use `std::optional` if you're afraid of using explicit sentinel values, such as `max_int`.

ivg replied:
it looks like that `optional` was added only in std+17, so you can't use it. So either use a sentinel or a boolean flag.

ivg commented on lib/bap_llvm/llvm_macho_loader.hpp:
```diff
@@ -152,6 +137,38 @@ MACHO_SECTION_FIELD(reserved2, uint32_t)
+
+void image_info(const macho &obj, ogre_doc &s) {
+    if (!is_exec(obj))
+        s.raw_entry("(entry 0)");
```
why is it 0, but not the lowest address? If it is a library then will this take into account the rebasing?

gitoleg replied:
it equals to zero because we use relative addresses. so after rebasing entry point will be equal to the lowest address

ivg commented on lib/bap_llvm/llvm_macho_loader.hpp:
```diff
@@ -152,6 +139,37 @@ MACHO_SECTION_FIELD(reserved2, uint32_t)
+            return;
+        }
+    }
+    uint64_t entry = INT_MAX;
```
The `INT_MAX` is not the maximum value of the `uint64_t` type, it is 

```
#include <limit>
std::numeric_limits<uint64_t>::max();
```

gitoleg commented:
So, finally,  I wrote `swap`  for the `relocation_info` struct and it is really better (and correct) than it was earler. Also added few comments for future readers.

ivg commented on lib/bap_llvm/llvm_macho_loader.hpp:
```diff
@@ -295,24 +311,40 @@ void iterate_macho_commands(const macho &obj, ogre_doc &s) {
-        const MachO::relocation_info *rel = reinterpret_cast<const MachO::relocation_info *>(ptr + next);
-        if (rel->r_address & MachO::R_SCATTERED) {
+        MachO::relocation_info rel = get_rel(obj, next);
+        if (rel.r_address & MachO::R_SCATTERED) {
```
`rel.r_address & MachO::R_SCATTERED` doesn't imply that the relocation `rel` is scattered, it also depends on the architecture, e.g., for x86_64 and arm64 the relocation would be an immediate value, i.e., 64 bit, so this iteration will go off track in case of 64 bit system. If needed I can give the same testing binary for the x86_64 arch.

There is a `isRelocationScattered` member-function of the `MachOObjectFile` class that can tell you for sure if it scattered or now. Is there any _good_ reason why you are not using this function? Also there are plenty of relocation related member functions that apparently takes the endianness into account. Why aren't you using them?

ivg commented on lib/bap_llvm/llvm_macho_loader.hpp:
```diff
@@ -336,13 +368,17 @@ void indirect_symbols(const macho &obj, const MachO::dysymtab_command &dlc, ogre
+
+                uint64_t index = n + j;
+                uint64_t offset = dlc.indirectsymoff + index * sizeof(uint32_t);
+                if (offset > obj_size)
```
what if `offset` is equal to `obj_size`?

ivg commented on lib/bap_llvm/llvm_macho_loader.hpp:
```diff
@@ -351,13 +387,19 @@ void indirect_symbols(const macho &obj, const MachO::dysymtab_command &dlc, ogre
 
+// This part is mainly for bundles, since it's not enough just to iterate over
+// sections in order to get relocations, e.g. like we do it for MH_OBJECT files.
+// We are going to take a look only to indirect symbols and external relocations.
```
s/to/for

ivg commented on lib/bap_llvm/llvm_macho_loader.hpp:
```diff
@@ -351,13 +387,19 @@ void indirect_symbols(const macho &obj, const MachO::dysymtab_command &dlc, ogre
+// sections in order to get relocations, e.g. like we do it for MH_OBJECT files.
+// We are going to take a look only to indirect symbols and external relocations.
+// Speaking about local relocations references from dyn sym tab, even modern (6.0)
+// does not perform a lot of job about them, so let's do the same. One cat take a look at
```
s/cat/can

ivg commented on lib/bap_llvm/llvm_macho_loader.hpp:
```diff
@@ -295,24 +311,40 @@ void iterate_macho_commands(const macho &obj, ogre_doc &s) {
 
 void indirect_symbols(const macho &obj, const MachO::dysymtab_command &dlc, ogre_doc &s) {
-
+    auto base = image_base(obj);
```
from here and below there is a lot of voluntarism with the handling  of the raw `base` pointer. Please, use some smart pointer, that will bound check every dereference. If you don't find any in C++ stdlib or LLVM, then implement your own.

gitoleg commented:
I will try to avoid to use the word `finally`) but anyway I like this implementation.

### Summary
I added more comments, and really hope they will be helpful.  Also added more guards and checks, so bug `shall not pass`, and refactored and simplified the code. There is only one place when we work directly with a raw pointer to object data, and I guarded it with checks too. 

### Relocations structures and endianess.
It turned out, that we should not use that `relocation_info` structure at all. I wrote a detailed explanation directly in code, but briefly - there are not any easy ways to read bit fields from it. We have to use masks, swaps, shifts. And actually, everything is already done in `llvm`, so we just need to use it. 

In the end, I checked everything on different files - little and big endian, and now I can say that it works like we expect.

ivg commented on lib/bap_llvm/llvm_macho_loader.hpp:
```diff
@@ -295,69 +313,136 @@ void iterate_macho_commands(const macho &obj, ogre_doc &s) {
-            next = next + sizeof(MachO::relocation_info);
-        }
+        auto rel = get_rel(obj, off + i * RELOCATION_ENTRY_SIZE);
+        if (rel && obj.getPlainRelocationExternal(*rel))
```
the `getPlainRelocationExternal` function depends on the endiannes, however, you already swapped the relocation bytes. Do you think that it is correct? I would suggest you to re-implement those LLVM functions and do not rely on them. Looks like this part of LLVM is buggy and was never tested, probably it is just easier to reimplement this stuff here.

ivg commented on lib/bap_llvm/llvm_macho_loader.hpp:
```diff
@@ -295,69 +313,136 @@ void iterate_macho_commands(const macho &obj, ogre_doc &s) {
+    // runtime LLVM_ERROR
+    uint64_t offset = dlc.indirectsymoff + index * sizeof(uint32_t);
+    symbol_iterator s = prim::end_symbols(obj);
+    if (offset < obj_size)
```
again, you can't compare offsets and sizes.

ivg commented on lib/bap_llvm/llvm_macho_loader.hpp:
```diff
@@ -295,69 +313,136 @@ void iterate_macho_commands(const macho &obj, ogre_doc &s) {
+    uint64_t offset = dlc.indirectsymoff + index * sizeof(uint32_t);
+    symbol_iterator s = prim::end_symbols(obj);
+    if (offset < obj_size)
+        s = get_symbol(obj, obj.getIndirectSymbolTableEntry(dlc, index));
```
instead of the pattern:
```
x = invalid;
if (valid(s)) {
  x = make(s);
  return x;
}
return x;
```

you can use much more safe and readable pattern:
```
if (valid(s)) 
    return make(s);
else
   return invalid;
```

This will make your code easier to understand (you do not need to track variables and their flow) and sustainable to changes (the invalid value doesn't roam in the scope waiting to be dereferenced).

ivg commented on lib/bap_llvm/llvm_macho_loader.hpp:
```diff
@@ -295,69 +313,136 @@ void iterate_macho_commands(const macho &obj, ogre_doc &s) {
+}
+
+// returns stride for indirect symbols entries
+uint32_t indirect_symbols_stride(const macho &obj, const SectionRef &sec) {
```
an easier to understand and less error prone version:
```C++
uint32_t indirect_symbols_stride(const macho &obj, const SectionRef &sec) {
    if (section_type(obj, sec) == MachO::S_SYMBOL_STUBS)
        return section_reserved2(obj, sec);
    else
        return obj.is64Bit() ? 8 : 4;
}
```

ivg commented on lib/bap_llvm/llvm_macho_loader.hpp:
```diff
@@ -295,69 +313,136 @@ void iterate_macho_commands(const macho &obj, ogre_doc &s) {
+// true for symbol stubs or symbol pointers sections
+bool contains_sym_stubs(const macho &obj, const SectionRef &sec) {
+    auto typ = section_type(obj, sec);
+    return
```
you do not need to put parentheses around the return expression. You can also use switch/case

```c++
switch (section_type(obj, sec)) {
  case X:
  case Y:  
  case Z:   return true;
  default:  return false;
}
```

That's is very close to OCaml's match.

ivg commented on lib/bap_llvm/llvm_macho_loader.hpp:
```diff
@@ -376,7 +461,13 @@ void symbols(const macho &obj, ogre_doc &s) {
     symbols(obj, sizes, s);
 }
 
+// It's safe to call getSymtabLoadCommand without any checks,
```
is it true for all versions? I remember that somewhere in LLVM 3.4 we were getting a segmentation fault when we were asking for a symtab. It might be in ELF backend though.

gitoleg replied:
it's true for llvm versions >= 3.8

ivg replied:
OK, for the record, in case of 3.4 this function will not be used

ivg commented on lib/bap_llvm/llvm_macho_loader.hpp:
```diff
@@ -295,70 +329,122 @@ void iterate_macho_commands(const macho &obj, ogre_doc &s) {
+
+bool is_rel_extern(const macho &obj, const MachO::any_relocation_info &rel) {
+    if (obj.isLittleEndian())
+        return (rel.r_word1 >> 27) & 1;
```
bless me father for I have sinned :)
